src/garage/torch/algos/maml/maml.py

```diff
@@ -0,0 +1,478 @@
+        meta_batch_size (int): Number of tasks sampled per batch.
+        inner_lr (double): Adaptation learning rate.
+        num_grad_updates (int): Number of adaptation gradient steps.
+        inner_algo (garage.torch.algos.VPG): The inner algorithm used for
+            computing loss.
+
+    """
+
+    def __init__(self,
+                 inner_algo,
+                 env,
+                 policy,
+                 baseline,
+                 meta_optimizer,
+                 meta_batch_size=40,
+                 inner_lr=0.1,
+                 num_grad_updates=1):
+        if policy.vectorized:
+            self.sampler_cls = OnPolicyVectorizedSampler
+        else:
+            self.sampler_cls = BatchSampler
+
+        self.max_path_length = inner_algo.max_path_length
+        self._policy = policy
+        self._env = env
+        self._baseline = baseline
+        self._num_grad_updates = num_grad_updates
+        self._meta_batch_size = meta_batch_size
+        self._inner_algo = inner_algo
+        self._inner_optimizer = DiffSGD(self._policy, lr=inner_lr)
+        self._meta_optimizer = make_optimizer(meta_optimizer,
+                                              policy,
+                                              lr=_Default(1e-3),
+                                              eps=_Default(1e-5))
+        self._episode_reward_mean = [
+            collections.deque(maxlen=100) for _ in range(num_grad_updates + 1)
+        ]
+
+    def train(self, runner):
+        """Obtain samples and start training for each epoch.
+
+        Args:
+            runner (LocalRunner): LocalRunner is passed to give algorithm
+                the access to runner.step_epochs(), which provides services
+                such as snapshotting and sampler control.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        last_return = None
+
+        for _ in runner.step_epochs():
+            all_samples, all_params = self._obtain_samples(runner)
+            tabular.record('TotalEnvSteps', runner.total_env_steps)
+            last_return = self.train_once(runner, all_samples, all_params)
+            runner.step_itr += 1
+
+        return last_return
+
+    def train_once(self, itr, all_samples, all_params):
+        """Train the algorithm once.
+
+        Args:
+            itr (int): Iteration number.
+            all_samples (list[list[MAMLTrajectoryBatch]]): A two
+                dimensional list of MAMLTrajectoryBatch of size
+                [meta_batch_size * (num_grad_updates + 1)]
+            all_params (list[dict]): A list of named parameter dictionaries.
+                Each dictionary contains key value pair of names (str) and
+                parameters (torch.Tensor).
+
+        Returns:
+            float: Average return.
+
+        """
+        old_theta = dict(self._policy.named_parameters())
+
+        kl_before = self._compute_kl_constraint(itr,
```

Comment:
how expensive is computing these KLs and losses before/after? perhaps we can try to capture them in the optimizer or turn it off if it's expensive?